juju/apiclient.py

```diff
@@ -17,10 +25,250 @@
+
+    If the Facade instances are being constructed through the Connection
+    instance, then the more specialized UnknownFacade and
+    FacadeVersionNotSupported errors will be raised, however if a Facade
+    instance is created through another method, the NotImplementedError will
+    be raised in the cases of bad facade names or versions.
+
+    """
+    def __init__(self, *args):
+        super(NotImplementedError, self).__init__(*args)
+
+
+
+def new_error(response):
+    """Constructs a specific exception type based on the ErrorCode."""
+    err_code = response.get("ErrorCode")
+    message = response.get("Error")
+    if err_code == "not implemented":
+        return NotImplementedError(err_code, message)
+    return ServerError(err_code, message)
+
+
+class Facade(object):
+    """The Facade instances provide convenient syntactic sugar for API calls.
+
+    Facade instances are constructed with a Connection, and facade name and
+    version. Any resulting method calls to a facade are translated into RPC
+    calls on the connection instance.
+
+    For example, calling the 'FullStatus' method on the client facade:
+
+        >>> client = Facade(connection, "Client")
+        >>> result = client.FullStatus()
+
+    This ends up being the same as:
+
+        >>> connection._rpc("Client", "FullStatus")
+
+    """
+
+    def __init__(self, connection, name, version=None):
+        self.connection = connection
+        self.name = name
+        self.version = version
+
+    def __getattribute__(self, attr):
+        try:
+            return object.__getattribute__(self, attr)
+        except AttributeError:
+            return functools.partial(self.connection._rpc, self.name, attr, version=self.version)
+
+
+class Connection(object):
+    """A connection represents an open, authenticated API connection."""
+    _request_id = 0
+    _upgrade_retry_count = 60
+    _upgrade_retry_delay_secs = 1
+
+    def __init__(self, address, cacert, auth_tag, password, nonce="", env_uuid=""):
+        self._cert_file = self._write_cert(cacert)
+        endpoint = self._endpoint(address, env_uuid)
+        cert_path = self._cert_file.name
+        self._connection = self._connect(endpoint, cert_path)
+        self._info = self._authenticate(auth_tag, password, nonce)
+        self._generate_facades()
+
+    def _authenticate(self, auth_tag, credentials, nonce):
+        # Start with version 2 of admin facade and work our way back.
+        for version in (2, 1, 0):
+            try:
+                self._auth_creds = self._login_args(version, auth_tag, credentials, nonce)
+                return self._rpc("Admin", "Login", self._auth_creds, version=version)
+            except NotImplementedError:
+                # do nothing and try the previous version
+                pass
+        raise RuntimeError("unexpected missing login command")
+
+    @staticmethod
+    def _login_args(version, auth_tag, credentials, nonce):
+        if version == 0:
```

Comment:
To avoid repetition:
```python
if version == 0:
    args = {"AuthTag": auth_tag, "Password": credentials}
else:
     # Otherwise, use the newer login structure.
    args = {"auth-tag": auth_tag, "credentials": credentials}
if nonce is not None:
    args['Nonce'] = nonce
return args
```